### PR TITLE
Fixed retrieving of incidence rates with execution info

### PR DESCRIPTION
**Pull request by aklochkova**
Fixes https://github.com/OHDSI/WebAPI/issues/677

### CODE REVIEW

File: src/main/java/org/ohdsi/webapi/ircalc/IncidenceRateAnalysisRepository.java [hunk 1]
```diff
@@ -25,6 +25,8 @@
 public interface IncidenceRateAnalysisRepository extends CrudRepository<IncidenceRateAnalysis, Integer> {
   
   @Query("select ira from IncidenceRateAnalysis AS ira LEFT JOIN FETCH ira.details as d")          
-  public Iterable<IncidenceRateAnalysis> findAll();
-  
+  Iterable<IncidenceRateAnalysis> findAll();
+
+  @Query("SELECT ira FROM IncidenceRateAnalysis AS ira JOIN FETCH ira.executionInfoList WHERE ira.id = ?1")
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/ircalc/IncidenceRateAnalysisRepository.java]: Why not `LEFT JOIN`? As soon as you query for IR w/o executions - this returns null. And then you get NPE at `IRAnalysisService.java` line 494

**Comment by pavgra** [comment]
IRs fail to be displayed also when a DS used for generation was deleted:
image

File: src/main/java/org/ohdsi/webapi/service/SourceService.java [hunk 2]
```diff
@@ -280,8 +284,10 @@ public Response delete(@PathParam("sourceId") Integer sourceId) throws Exception
     Source source = sourceRepository.findBySourceId(sourceId);
     if (source != null) {
       final String sourceKey = source.getSourceKey();
+      List<ExecutionInfo> executions = irExecutionInfoRepository.findBySource(source);
+      irExecutionInfoRepository.delete(executions);
       sourceRepository.delete(source);
-        publisher.publishEvent(new DeleteDataSourceEvent(this, sourceId, source.getSourceName()));
+      publisher.publishEvent(new DeleteDataSourceEvent(this, sourceId, source.getSourceName()));
```
Comment (chrisknoll) [inline src/main/java/org/ohdsi/webapi/service/SourceService.java]: This change seems out of place inside a PR that deals with retrieving incidence rate execution info?

File: src/main/java/org/ohdsi/webapi/service/SourceService.java [hunk 3]
```diff
@@ -59,6 +61,8 @@
   private Environment env;
   @Autowired
   private ApplicationEventPublisher publisher;
+  @Autowired
+  private IRExecutionInfoRepository irExecutionInfoRepository;
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/service/SourceService.java]: Where is it used?

File: src/main/java/org/ohdsi/webapi/service/SourceService.java [hunk 4]
```diff
@@ -281,7 +285,7 @@ public Response delete(@PathParam("sourceId") Integer sourceId) throws Exception
     if (source != null) {
       final String sourceKey = source.getSourceKey();
       sourceRepository.delete(source);
-        publisher.publishEvent(new DeleteDataSourceEvent(this, sourceId, source.getSourceName()));
+      publisher.publishEvent(new DeleteDataSourceEvent(this, sourceId, source.getSourceName()));
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/service/SourceService.java]: Dummy diff

File: src/main/java/org/ohdsi/webapi/service/IRAnalysisService.java [hunk 5]
```diff
@@ -104,6 +105,7 @@
  * @author Chris Knoll <cknoll@ohdsi.org>
  */
 @Path("/ir/")
+@Singleton
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/service/IRAnalysisService.java]: Where is the Service created more than once? Maybe we should fix cause but not consequences?

**Comment by chrisknoll** [comment]
Hi,
 This PR is changing to the structore of Service -> ServiceImpl. 

We haven't talked about this project structure, and I am opposed to this needless layer.  We need to review what was done on CC and pathways to determine which elements we liked, which elements we didn't like, but we are not going to push that structure to every other part of the application.

This is something we can review on the WebAPI call.

**Comment by chrisknoll** [comment]
Hi, @aklochkova ,
  I need to step-back from my previous objection:  @pavgra walked me through the Spring core code that is having the challenge with handling javax.ws (Jersey) endpoints and the interop with spring:  Jersey and Spring both have their own dependency injections, and in this particular case the IRAnalyssisService extending an interface is making the mechanism in Spring create an instance of the service per request, which is leading ot the post-create call being invoked multiple times, which is leading to the incomplete job cleanup being fired multiple times.  All this is bad, and the simplest fix was to extract all the Jersey annotations out into a single interface so that the service is not re-initialized each request.


So, this change isn't to push the controller-service-serviceImpl pattern, but rather to just fix a problem that was introdued when adding the 'GeneratesNotifiation' as an implemented interface...fine.

Note for 3.0:  the cleanup of the incomplete services should probably not have been implemented as a function of each service. Rather there should have been a single bean that gets invoked once-per-startup of WebAPI that performs the cleanup across all services.  This separates the concern from 'Do this activity at startup' of this initialization bean from 'Handle requests from clients' that is the job of each of the services.

This last point isn't placing blame at anyone, just wanted to remember this idea for when we do the larger-scale refactoring for webAPI 3.0.

**Pull request merged**